### PR TITLE
Ensure selected site exist before accessing it when loading stats

### LINKED ISSUE #5688: Crash on GetStats use case
Sentry issue ID: [WOOCOMMERCE-ANDROID-2Y7](https://sentry.io/organizations/a8c/issues/2892184480/?referrer=github_integration)

Crash when trying to retrieve selected site: 
```
java.lang.IllegalStateException: SelectedSite.get() was accessed before being initialized - siteId -1.
Consider calling selectedSite.exists() to ensure site exists prior to calling selectedSite.get().
    at com.woocommerce.android.tools.SelectedSite.get(SelectedSite.kt:48)
    at com.woocommerce.android.ui.mystore.data.StatsRepository$fetchRevenueStats$2.invokeSuspend(StatsRepository.kt:38)
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
    at kotlinx.coroutines.UndispatchedCoroutine.afterResume(CoroutineContext.kt:142)
    at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:102)
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
    at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
    at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

### PR DESCRIPTION
**Pull request by JorgeMucientes**
⚠️ DO NOT MERGE

Closes: #5688 
Sentry issue ID: [WOOCOMMERCE-ANDROID-2Y7](https://sentry.io/organizations/a8c/issues/2892184480/?project=1459556&referrer=github_integration)

### Description
I was unable to reproduce this crash. I believe it might be some kind of race condition after selecting the site on Login. Basically, MainActiivty has a check in the `onCreate` fun that ensures a site is selected before loading any Fragment. So, by the time MyStoreFragment is loaded and `GetStats` use case is invoked, selectedSite should always exist. 

The stacktrace is pretty clear about the issue
```
java.lang.IllegalStateException: SelectedSite.get() was accessed before being initialized - siteId -1.
Consider calling selectedSite.exists() to ensure site exists prior to calling selectedSite.get().
    at com.woocommerce.android.tools.SelectedSite.get(SelectedSite.kt:48)
    at com.woocommerce.android.ui.mystore.data.StatsRepository$fetchRevenueStats$2.invokeSuspend(StatsRepository.kt:38)
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
    at kotlinx.coroutines.UndispatchedCoroutine.afterResume(CoroutineContext.kt:142)
    at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:102)
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
    at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
    at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

The flow from the crash show that `SitePickerActivity` is never opened after login. Maybe the user is trying to access the app and does not have any existing site? 

Screenshot 2022-01-28 at 09 16 44

### Testing instructions
I could not reproduce this so there is not testing instructions. What I did is force the scenario were selectedSite is null and we call `GetStats`. With the new check and empty state will be shown:

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/data/StatsRepository.kt [hunk 1]
```diff
@@ -30,42 +30,52 @@ class StatsRepository @Inject constructor(
...
     suspend fun fetchVisitorStats(granularity: StatsGranularity, forced: Boolean): Flow<Result<Map<String, Int>>> =
         flow {
-            val visitsPayload = FetchNewVisitorStatsPayload(selectedSite.get(), granularity, forced)
-            val result = wcStatsStore.fetchNewVisitorStats(visitsPayload)
-            if (!result.isError) {
-                val visitorStats = wcStatsStore.getNewVisitorStats(
-                    selectedSite.get(), result.granularity, result.quantity, result.date, result.isCustomField
-                )
-                emit(Result.success(visitorStats))
+            if (selectedSite.exists()) {
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/data/StatsRepository.kt:60]: This method was not generating any crash, but I believe the same issue could happen, as both are called at the same time.

**Review by malinajirka — COMMENTED** [review]
Thanks for looking into this and for your attempts to reproduce the crash @JorgeMucientes! 

I think the solution will work, but tbh I'm not sure if it's worth it. Stats screen should never be shown without a valid/selected site. If it is shown there is a bug in the app. This solution merely hides the bug, does not actually solve it. 

What I'm worried about is that we end up with tens/hunders of `exists()` checks all over the codebase, even when only a few of them will be actually necessary. This might make the codebase quite hard to maintain in the long run.

What I also noticed is that it seems the `userId` seems to be almost always `0` (only one time it's non-empty). So it seems like we are trying to open stats before the user is actually logged in - or perhaps before the login process completes, might be a race condition as you mentioned?

Having said all that, I don't mind merging it, but I think it might be worth asking others what they think about it. It seems like a deeper issue. Wdyt?

cc @anitaa1990 @hichamboushaba @wzieba

**Comment by malinajirka** [comment]
I noticed two more things from the logs - I'm not sure how it is related yet, but perhaps someone else will have an explanation.
1. There are two instances of LoginActivity living in parallel.
2. When the app crashes, it almost seems like neither MainActivity nor AppSettingsActivity even exist - they were destroyed.

**Comment by JorgeMucientes** [comment]
Thanks for checking this @malinajirka. I agree with your points. I though this fix could mitigate the impact on UX, considering it is a race condition, with the fix at least the app does not crash and the moment you switch tab on stats screen, or swipe to refresh the stats will be loaded again. 
But on the other hand, I'm not 100% sure is a race condition, and as you mention this fix is at the cost of hiding the real bug and adding unnecessary `exist()` checks. 

Let's not merge this and let me try to investigate further to see if I can find the root cause.

**Comment by hichamboushaba** [comment]
I have a theory, but I need to confirm it, I'll lay it down before testing it:

In MainActivity, the first thing we do is we attach the navgraph [here](https://github.com/woocommerce/woocommerce-android/blob/a797900c552255bd4d9b20731a19e24c14fe6a93/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt#L199-L205), then we check if the user is logged in or not [here](https://github.com/woocommerce/woocommerce-android/blob/a797900c552255bd4d9b20731a19e24c14fe6a93/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt#L208-L211), as the Activity navigation actions are async, they `MyStoreFragment` may reach the state created before we finish `MainActivity`, and as we are loading stats in the `MyStoreViewModel`'s constructor now, we may start loading them too even if the user is not logged in.

If this theory is confirmed, I think a better fix would be to avoid attaching the `navgraph` if the user is not logged in, by changing the order of the calls.

**Comment by hichamboushaba** [comment]
The code I linked above is not responsible for attaching the `navgraph`, it's actually attached in the XML, and with the `startDestination`, it doesn't depend on my code above.

This doesn't invalidate the theory, just the fix would be different, as now we need to avoid specifying the `navgraph` in the XML, and doing it in the code behind after checking the user is logged in.

I was able to replicate a very similar crash by delaying the `finish` to `onResume` using this patch:
```
Index: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt
--- a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt	(revision 843f990c92358a8682856dd108c762c3d4ed2456)
+++ b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt	(date 1643367470620)
@@ -7,6 +7,7 @@
 import android.content.res.Resources.Theme
 import android.net.Uri
 import android.os.Bundle
+import android.os.Handler
 import android.view.Menu
 import android.view.MenuItem
 import android.view.View
@@ -254,6 +255,8 @@
 
     override fun onResume() {
         super.onResume()
+        finish()
+        return
         AnalyticsTracker.trackViewShown(this)
 
         updateReviewsBadge()
@@ -578,7 +581,7 @@
         val intent = Intent(this, LoginActivity::class.java)
         LoginMode.WOO_LOGIN_MODE.putInto(intent)
         startActivityForResult(intent, RequestCodes.ADD_ACCOUNT)
-        finish()
+        //finish()
     }
 
     /**
```
But I wouldn't really take this as a confirmation.

**Comment by JorgeMucientes** [comment]
Thanks for checking this out @hichamboushaba, that theory makes sense, but I've done a quick test to try to reproduce the crash by a adding a Thread.sleep() in between setting the setting of navgrapth and checking for user logged in state: 
```kotlin 
        navHostFragment.childFragmentManager.registerFragmentLifecycleCallbacks(fragmentLifecycleObserver, false)

        binding.bottomNav.init(navController, this)

        Thread.sleep(3000)
        // Verify authenticated session
        if (!presenter.userIsLoggedIn()) {
            showLoginScreen()
            return
        }
```

Although not sure if this is a proper test, as blocking the main thread for 3s also will avoid MyStoreFragment from being instantiated, since that is done from the main thread too. Any idea on how to try to reproduce the bug to make sure this is the root cause of the issue?

Another doubt I have is that I don't know exactly when that potential navigation to `MyStoreFragment` is really happening. Afaik this code
```kotlin 
        val navHostFragment = supportFragmentManager.findFragmentById(R.id.nav_host_fragment_main) as NavHostFragment
        navController = navHostFragment.navController
        navController.addOnDestinationChangedListener(this@MainActivity)

        navHostFragment.childFragmentManager.registerFragmentLifecycleCallbacks(fragmentLifecycleObserver, false)

        binding.bottomNav.init(navController, this)
```
Is not responsible for navigating to `MyStoreFragment` but to `init` the bottomNav right? The code that should trigger the navigation to MyStoreFragment should happen anytime after `setContentView(binding.root)` is called? As that is the moment were we are setting the `activity_main.xml` that sets the `nav_graph_main` to be used.

**Comment by hichamboushaba** [comment]
@JorgeMucientes we commented on the same time 😄

**Comment by JorgeMucientes** [comment]
Indeed!! I think we are saying more or less the same, plus you answered some of the doubts I had. I think we can still keep specifying the navgraph in the XM. Maybe we can just check the login state before setting the activity content. 
```kotlin
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        presenter.takeView(this)
        // Verify authenticated session
        if (!presenter.userIsLoggedIn()) {
            showLoginScreen()
            return
        }

        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
...
```
Going to test if I can reproduce the bug with the patch you suggested and then test possible solutions

**Comment by hichamboushaba** [comment]
> I think we can still keep specifying the navgraph in the XM. Maybe we can just check the login state before setting the activity content.

Yes, this would work too, thanks @JorgeMucientes

**Comment by JorgeMucientes** [comment]
I'm going to close this PR @malinajirka @hichamboushaba. I've opened this other PR: #5728 were the proper fix is applied. The fix will be merged into `release/8.4` in the new PR, as it looks safe. 

Thanks a lot for your input :)

**Pull request closed without merging**